experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -89,112 +93,117 @@ LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
+      : valueForSpecTag(valueForSpecTag), branchOp(branchOp),
+        branchDirection(branchDirection) {}
+};
 
-// This recursive function traverses the IR along a marked path and creates
-// a control path by replicating the branches it finds in the way. It stops
-// at commits and connects them to the newly created path with value
-// ctrlSignal
-void HandshakeSpeculationPass::routeCommitControl(
-    llvm::DenseSet<Operation *> &markedPath, Value ctrlSignal,
-    OpOperand &currOpOperand) {
+// This function traverses the IR and creates a control path by replicating the
+// branches it finds in the way. It stops at commits and connects them to the
+// newly created path with value ctrlSignal
+static void
+routeCommitControlRecursive(MLIRContext *ctx, SpeculatorOp &specOp,
+                            llvm::DenseSet<Operation *> &arrived,
+                            OpOperand &currOpOperand,
+                            std::vector<BranchTracingItem> &branchTrace) {
   Operation *currOp = currOpOperand.getOwner();
-  // End traversal if currOp is not in the marked path to commits
-  if (!markedPath.contains(currOp))
+
+  // End traversal if currOpOperand is already arrived
+  if (arrived.contains(currOp))
     return;
+  arrived.insert(currOp);
 
-  // Remove operation from the set to avoid visiting it twice
-  markedPath.erase(currOp);
+  // We assume there is a direct path from the speculator to all commits, and so
+  // traversal ends if we reach a save-commit or a speculator. See detailed
+  // documentation for full explanation of the speculative region and this
+  // assumption.
+  if (isa<handshake::SpeculatorOp>(currOp))
+    return;
+  if (isa<handshake::SpecSaveCommitOp>(currOp))
+    return;
 
   if (auto commitOp = dyn_cast<handshake::SpecCommitOp>(currOp)) {
+    // We replicate branches only if the traversal reaches a commit.
+    // Because sometimes a path of branches does not reach a commit unit.
+    Value ctrlSignal = specOp.getCommitCtrl();
+    for (auto [valueForSpecTag, branchOp, branchDirection] : branchTrace) {
+      // Replicate a branch in the control path and use new control signal.
+      // To do so, a structure of two connected branches is created.
+      // A speculating branch first discards the condition in case that
+      // the data is not speculative. In case it is speculative, a new branch
+      // is created that replicates the current branch.
+
+      OpBuilder builder(ctx);
+      builder.setInsertionPointAfterValue(ctrlSignal);
+
+      // The speculating branch will discard the branch's condition token if
+      // the branch output is non-speculative. Speculative tag of the token is
+      // currently implicit, so the branch input itself is used at the IR
+      // level.
+      auto branchDiscardNonSpec =
+          builder.create<handshake::SpeculatingBranchOp>(
+              branchOp.getLoc(), valueForSpecTag /* specTag */,
+              branchOp.getConditionOperand());
```

Comment:
We definitely want to use the condition twice here, I don't know materialization super well but don't we just call that pass again after the speculation pass?

Reply:
Materialization is just inserting a fork if a value is used multiple times (a channel cannot have multiple consumers). I think it throws an error if this is violated at the end of the pass.

Reply:
We're now running the SpeculationPass with the --handshake-materialize argument. It seems this pass has several cases where the same value is referenced by different operands, so I think it makes sense to perform materialization after this pass.

Reply:
Maybe it makes more sense to always run materialize pass after speculation stuff (you can call other passes in your pass), here is how I did it in some other places:
https://github.com/EPFL-LAP/dynamatic/blob/815081c554b974debfefba96177a949e18e8c844/experimental/lib/Transforms/ResourceSharing/Crush.cpp#L397-L403

Reply:
I still don't understand the point you're trying to make- we do not need to worry about how many times a value is used, because we call the pass with the --handshake_materialize option, which takes care of all of this of for us

Reply:
I agree. I'll do that, is a different PR for this better?

Reply:
I would go so far as to say addPass is not at all intended to be used like this-
![image](https://github.com/user-attachments/assets/84989d13-ab66-47dd-a11a-10423e7885ad)
All the passes that are run should be listed explicitly, and should show the pass order, in compile.sh

someone should be able to read the compile.sh and know the exact pass order, and using addPass hides the fact that a pass is being used

Reply:
I am convinced now. I previously thought every dynamatic pass should maintain materialization; but it seems not so (if it is required then the pass just checks at the beginning

I already planned to remove this addPass thing at some point

Reply:
nice!